src/controller/worker.rs

```diff
@@ -89,91 +79,89 @@ impl Worker {
         block_range: Option<(u64, u64)>,
         client_id: Option<PeerId>,
     ) -> QueryResult {
-        let (resp_tx, resp_rx) = oneshot::channel();
-        match self.queries_tx.try_send(QueryTask {
-            dataset,
-            query_str,
-            block_range,
-            client_id,
-            response_sender: resp_tx,
-        }) {
-            Err(mpsc::error::TrySendError::Full(_)) => {
-                return Err(QueryError::ServiceOverloaded);
-            }
-            Err(mpsc::error::TrySendError::Closed(_)) => {
-                panic!("Query subscriber dropped");
-            }
-            Ok(_) => {
-                metrics::PENDING_QUERIES.inc();
-            }
-        };
-        resp_rx
-            .await
-            .expect("Query processor didn't produce a result")
+        let before = self.queries_running.fetch_add(1, Ordering::SeqCst);
+        let counter_guard = scopeguard::guard((), |_| {
+            let before = self.queries_running.fetch_sub(1, Ordering::SeqCst);
+            metrics::RUNNING_QUERIES.set(before as i64 - 1);
```

Comment:
Here we have a data race. `inc()` / `dec()` should be used.

In JS land prometheus metrics can be pulled. Can we move this metric code away from this component?

Reply:
> In JS land prometheus metrics can be pulled. Can we move this metric code away from this component?

I don't think so. How can a metrics component find the Worker instance? It would require adding some globals which would complicate the code even more. Could you provide an example how it is done in JS?

Metrics are pulled on the protocol level, so you can do arbitrary updates just when `/metrics` got called, but the `prometheus_client` library suggests to use these global counters and update them from the code.

Reply:
> Could you provide an example how it is done in JS?

```rust
fn main() {
  let worker = Worker::new();
  metrics::register("running_queries", || worker.running_queries())
}
```
```

Reply:
Yeah, this pattern is strongly discouraged by Rust type system, unfortunately